### ISSUE TITLE
Functional middleware and listeners

### ISSUE THREAD

**Issue by pbsds** [kennethreitz/responder#150]
This is related to #58, but focuses on _middleware_.
Middlewares are nice to have when you for example want the same error handling across all your routes. It removes a lot of the noise you'd otherwise have to see in your codebase.
They can be used to handle 404 or 500, they could set certain headers/cookies or they could be used to enforce auth.

it could look something like this for responder:
```python
@api.middleware('before')
def pizzafy(req, resp):
    resp.headers['X-Pizza'] = 42

@api.middleware('before')
def strict_slashes(req, resp):
    if not req.url.path.endswith("/"):
        api.redirect(resp, req.url.copy_with(path=req.url.path+"/"))
    return True # one possible way to "respond early"

@api.middleware('after')
def pretty_404(request, response):
    if resp.status_code == status_codes.HTTP_404:
        resp.text = api.template('404.html')

```

There are multiple approaches taken by the other frameworks:
* [Sanic](https://sanic.readthedocs.io/en/latest/sanic/middleware.html)
* [molten](https://moltenframework.com/#functional-middleware)
* [aiohttp](https://docs.aiohttp.org/en/stable/web_advanced.html#middlewares)
* [flask](http://flask.pocoo.org/docs/1.0/api/#flask.Flask.before_request)

Sanic and Flask have in my opinion the easiest/prettiest ones to use, which are pretty fit for responder.
 
aiohttp's idea of registering middlewares as implicit decorators for all routes/handlers however is pretty powerful and i like the idea. I'd opt for implementing something in the same vein as aiohttp, but it does have the weakness of executing only after the route has been matched by responder, making it unable to rewrite the request url before matching.

I tried prototyping an ASGI middleware class to be used with `api.add_middlware`, which does something like this, but then I only got access to uvicorn's req and resp objects, not the ones in `responder.models`.

**Comment by kennethreitz** [comment]
I'm not a fan of this approach.

**Comment by pbsds** [comment]
This is meant as a discussion. Closing it gets us nowhere. 

If you don't like the changes this entails to the codebase or this kind of functionality at all, then please at least say so. If you don't like the decorator names, then renaming them should be simple. 

```python
@api.before_routing
def pizzafy(req, resp):
    resp.headers['X-Pizza'] = 42
```

**Comment by gigatexal** [comment]
So...where does that leave middlewares and this framework?

**Comment by chrisbrake** [comment]
Because it's based on starlette, use this method:
https://www.starlette.io/middleware/

**Comment by pbsds** [comment]
ASGI middleware is able to do some of the things middleware in other framworks is able to do, but you won't get access to responders reqeust and response objects. What you CAN get are uvicorns variants.

This for example, leaves out modifying the session cookie in an easy manner.

Its possible to work around this by applying a decorator to every route, but that is incredibly unwieldy.

**Comment by tomchristie** [comment]
We can build ASGI middleware classes that expose a request/response interface to the developer, but satisfy the ASGI interface.

That gives you the best of both worlds (interoperability, plus developer exerperience.)

I’ll be implementing that for Starlette hopefully sometime over the next week.

**Comment by tomchristie** [comment]
Example: https://github.com/encode/starlette/pull/155

That implementation gives you Starlette's `Request` and `Response` interfaces within the middleware, but could equally well be responder models. Provides the same kinda developer interface, but concrete implementations will be compatible with any ASGI framework, rather than strictly tied to one particular framework.

(If we're just a little bit careful about things like session interfaces we can likely make provide a degree of interoperability at that level too, without being tied to a specific session implementation, tho I've not considered that in depth yet)

**Comment by wshayes** [comment]
Is it possible to expose this Starlette approach (Example: encode/starlette#155) in responder?
